core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java

```diff
@@ -0,0 +1,149 @@
+
+    /** the current local checkpoint, i.e., all seqNo lower&lt;= this number have been completed */
+    volatile long checkpoint = -1;
+
+    /** the next available seqNo - used for seqNo generation */
+    volatile long nextSeqNo = 0;
+
+
+    public LocalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+        super(shardId, indexSettings);
+        bitArraysSize = indexSettings.getSettings().getAsInt(SETTINGS_BIT_ARRAYS_SIZE, DEFAULT_BIT_ARRAYS_SIZE);
+        if (bitArraysSize <= 0) {
+            throw new IllegalArgumentException("[" + SETTINGS_BIT_ARRAYS_SIZE + "] must be positive. got [" + bitArraysSize + "]");
+        }
+        processedSeqNo = new LinkedList<>();
+    }
+
+    /**
+     * issue the next sequence number
+     **/
+    public synchronized long generateSeqNo() {
+        return nextSeqNo++;
+    }
+
+    /**
+     * marks the processing of the given seqNo have been completed
+     **/
+    public synchronized void markSeqNoAsCompleted(long seqNo) {
+        // make sure we track highest seen seqNo
+        if (seqNo >= nextSeqNo) {
+            nextSeqNo = seqNo + 1;
+        }
+        if (seqNo <= checkpoint) {
+            // this is possible during recovery where we might replay an op that was also replicated
+            return;
+        }
+        FixedBitSet bitSet = getBitSetForSeqNo(seqNo);
+        int offset = seqNoToBitSetOffset(seqNo);
+        bitSet.set(offset);
+        if (seqNo == checkpoint + 1) {
+            updateCheckpoint();
+        }
+    }
+
+    /** get's the current check point */
+    public long getCheckpoint() {
+        return checkpoint;
+    }
+
+    /** get's the maximum seqno seen so far */
+    public long getMaxSeqNo() {
+        return nextSeqNo - 1;
+    }
+
+    /**
+     * moves the checkpoint to the last consecutively processed seqNo
+     * Note: this method assumes that the seqNo following the current checkpoint is processed.
+     */
+    private void updateCheckpoint() {
+        assert Thread.holdsLock(this);
+        assert checkpoint - firstSeqNoInProcessSeqNo < bitArraysSize : "checkpoint to firstSeqNoInProcessSeqNo is larger then a bit set";
+        assert getBitSetForSeqNo(checkpoint + 1).get(seqNoToBitSetOffset(checkpoint + 1)) : "updateCheckpoint is called but the bit following the checkpoint is not set";
+        assert getBitSetForSeqNo(checkpoint + 1) == processedSeqNo.getFirst() : "checkpoint + 1 doesn't point to the first bit set";
+        // keep it simple for now, get the checkpoint one by one. in the future we can optimize and read words
+        FixedBitSet current = processedSeqNo.getFirst();
+        do {
+            checkpoint++;
+            // the checkpoint always falls in the first bit set or just before. If it falls
+            // on the last bit of the current bit set, we can clean it.
+            if (checkpoint == firstSeqNoInProcessSeqNo + bitArraysSize - 1) {
+                processedSeqNo.pop();
```

Comment:
While `LinkedList#pop` is equivalent to `LinkedList#removeFirst`, since in other places `LinkedList#getFirst` and `LinkedList#peekFirst` is being invoked, it would be clearer if this was replaced by `LinkedList#removeFirst`.